opentasks-provider/src/main/java/org/dmfs/provider/tasks/processors/instances/TaskValueDelegate.java

```diff
@@ -0,0 +1,167 @@
+            task.set(TaskAdapter.ORIGINAL_INSTANCE_TIME, entityAdapter.valueOf(InstanceAdapter.INSTANCE_ORIGINAL_TIME));
+            task.unset(TaskAdapter.SYNC1);
+            task.unset(TaskAdapter.SYNC2);
+            task.unset(TaskAdapter.SYNC3);
+            task.unset(TaskAdapter.SYNC4);
+            task.unset(TaskAdapter.SYNC5);
+            task.unset(TaskAdapter.SYNC6);
+            task.unset(TaskAdapter.SYNC7);
+            task.unset(TaskAdapter.SYNC8);
+            task.unset(TaskAdapter.SYNC_ID);
+            task.unset(TaskAdapter.SYNC_VERSION);
+            // TODO: if this is the first instance (and maybe no other overrides exist), don't create an override but split the series into two tasks
+            mDelegate.insert(db, task, false);
+        }
+        else
+        {
+            // this is a non-recurring task or it's already an override, just delegate the update
+            mDelegate.update(db, taskAdapter, false);
+        }
+        return entityAdapter;
+    }
+
+
+    @Override
+    public void delete(SQLiteDatabase db, InstanceAdapter entityAdapter, boolean isSyncAdapter)
+    {
+        // deleted instances are converted to deleted tasks (for non-recurring tasks) or exdates (for recurring tasks).
+        TaskAdapter taskAdapter = entityAdapter.taskAdapter();
+
+        if (taskAdapter.valueOf(TaskAdapter.ORIGINAL_INSTANCE_ID) != null)
+        {
+            /* this is an override - we have to:
+             * - mark it deleted
+             * - add an exclusion to the master task
+             *
+             * TODO: if this is the first instance, consider moving the recurrence start instead of adding an exdate
+             */
+            long masterTaskId = taskAdapter.valueOf(TaskAdapter.ORIGINAL_INSTANCE_ID);
+            DateTime originalTime = entityAdapter.valueOf(InstanceAdapter.INSTANCE_ORIGINAL_TIME);
+
+            // delete the override
+            mDelegate.delete(db, taskAdapter, false);
+
+            // get the master and add an exdate
+            Cursor c = db.query(TaskDatabaseHelper.Tables.TASKS, null /* all */, TaskContract.Tasks._ID + "=" + masterTaskId, null, null, null, null);
+            try
+            {
+                if (c.moveToFirst())
+                {
+                    TaskAdapter masterTaskAdapter = new CursorContentValuesTaskAdapter(masterTaskId, c, new ContentValues());
+                    addExDate(masterTaskAdapter, originalTime);
+                    mDelegate.update(db, masterTaskAdapter, false);
+                }
+            }
+            finally
+            {
+                c.close();
+            }
+        }
+        else if (taskAdapter.isRecurring())
+        {
+            // TODO: if this is the first instance, consider moving the recurrence start instead of adding an exdate
+            addExDate(taskAdapter, entityAdapter.valueOf(InstanceAdapter.INSTANCE_ORIGINAL_TIME));
+            mDelegate.update(db, taskAdapter, false);
```

Comment:
Is updating `EXDATE` going to trigger a remove of that instance somehow?

Another related question: are the master task instance rows which have the RDATE/RRULE going to be exempt from actual queries, like what to show on the UI? And another instance row will be generated for the starting instance?
I am asking this because otherwise, if the first instance is the master itself, then how can we handle the case when that is removed from the recurrence set? (Which is what this `else if` branch is about if I understand correctly.)

Reply:
Yes, adding an `EXDATE` will result in removing the instance row.

All recurring instances which are not overridden will point to the master task. When you remove the first instance, an `EXDATE` is added and the instance row of the first instance is removed. Nothing else will happen. That reminds me, when all instances are removed we probably should remove the entire task.

Reply:
Ok, and regarding the query for the tasks shown in the app, we will exclude the master task row, right? Since we will have a separate instance row for the first instance.

Reply:
we don't explicitly exclude the master task. If there happens to be an override for every instance we don't show the master task data, but if some instances have no overrides, they will be joined with the master task (since `TASK_ID` still points there)

Reply:
Oh, I understand now, so the master task row in Task table doesn't even get a row in Instances table in itself. Only the actual enumerated instances and overrides. Makes sense, considering the 'Instances' name :) (don't know how I could imagine it otherwise)